javatests/dagger/functional/assisted/AssistedFactoryTest.java

```diff
@@ -0,0 +1,302 @@
+    @Inject
+    Dep1(Dep2 dep2, Dep3 dep3) {}
+  }
+
+  static final class Dep2 {
+    @Inject
+    Dep2(Dep3 dep3) {}
+  }
+
+  static final class Dep3 {
+    @Inject
+    Dep3(Dep4 dep4) {}
+  }
+
+  static final class Dep4 {
+    @Inject
+    Dep4() {}
+  }
+
+  // A base interface to test that factories can reference subclasses of the assisted parameter.
+  interface AssistedDep {}
+
+  static final class AssistedDep1 implements AssistedDep {}
+
+  static final class AssistedDep2 implements AssistedDep {}
+
+  static final class SimpleFoo {
+    private final AssistedDep assistedDep;
+
+    @AssistedInject
+    SimpleFoo(@Assisted AssistedDep assistedDep) {
+      this.assistedDep = assistedDep;
+    }
+
+    @AssistedFactory
+    interface Factory {
+      // Use different parameter names than Foo to make sure we're not assuming they're the same.
+      SimpleFoo createSimpleFoo(AssistedDep factoryAssistedDep);
+
+      // A no-op method to test static methods in assisted factories
+      static void staticMethod() {
+        return;
+      }
+
+      // A no-op method to test default methods in assisted factories
+      default void defaultMethod() {
+        return;
+      }
+    }
+  }
+
+  @AssistedFactory
+  interface SimpleFooFactory {
+    // Use different parameter names than Foo to make sure we're not assuming they're the same.
+    SimpleFoo createSimpleFoo(AssistedDep1 factoryAssistedDep1);
+
+    // A no-op method to test static methods in assisted factories
+    static void staticMethod() {
+      return;
+    }
+
+    // A no-op method to test default methods in assisted factories
+    default void defaultMethod() {
+      return;
+    }
+  }
+
+  @AssistedFactory
+  interface ExtendedSimpleFooFactory extends SimpleFooFactory {}
+
+  abstract static class BaseFoo {
+    @Inject Dep4 dep4;
+  }
+
+  static final class Foo extends BaseFoo {
+    private final Dep1 dep1;
+    private final Provider<Dep2> dep2Provider;
+    private final AssistedDep1 assistedDep1;
+    private final AssistedDep2 assistedDep2;
+    private final int assistedInt;
+    private final FooFactory factory;
+
+    @Inject Dep3 dep3;
+
+    @AssistedInject
+    Foo(
+        Dep1 dep1,
+        @Assisted AssistedDep1 assistedDep1,
+        Provider<Dep2> dep2Provider,
+        @Assisted AssistedDep2 assistedDep2,
+        @Assisted int assistedInt,
+        FooFactory factory) {
+      this.dep1 = dep1;
+      this.dep2Provider = dep2Provider;
+      this.assistedDep1 = assistedDep1;
+      this.assistedDep2 = assistedDep2;
+      this.assistedInt = assistedInt;
+      this.factory = factory;
+    }
+  }
+
+  @AssistedFactory
+  interface FooFactory {
+    // Use different parameter names than Foo to make sure we're not assuming they're the same.
+    Foo createFoo(
+        AssistedDep1 factoryAssistedDep1, AssistedDep2 factoryAssistedDep2, int factoryAssistedInt);
+  }
+
+  @AssistedFactory
+  abstract static class AbstractFooFactory {
```

Comment:
`abstract class` factories can have normal methods. There's a test for `static` and `default` methods on an interface. I would probably add `static` + normal methods for this also.

Reply:
Thanks, I can add that in a follow-up. (These pull requests are merged automatically on each internal submit, so we usually don't see these external comments on the pull request until after it's already accepted.)

Reply:
Yeah I know. Just wanted to give it a look over since it's a feature I'm all-too familiar with.